File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/template/concrete_support.dart
import 'abstract_class.dart';

class ContactFlow extends FormTemplate {
  @override
  Map<String, dynamic> validate(Map<String, dynamic> data) {
    if (data['subject'] == null || (data['subject'] as String).trim().isEmpty) {
      throw Exception('Subject is required');
    }
    if (data['message'] == null || (data['message'] as String).trim().isEmpty) {
      throw Exception('Message is required');
    }
    return {
      'subject': (data['subject'] as String).trim(),
      'message': (data['message'] as String).trim(),
      'priority': data['priority'] ?? 'normal',
    };
  }

  @override
  Future<dynamic> send(Map<String, dynamic> cleanData) async {
    print('[Contact] sending ticket: $cleanData');
    return {
      'ticketId': 'TICKET-' + DateTime.now().millisecondsSinceEpoch.toString(),
      'status': 'open',
    };
  }

  @override
  Future<void> afterSuccess(result) async {
    print('[Contact] notify support about ${result['ticketId']}');
  }

  @override
  String successMessage(result) {
    return 'Ticket ${result['ticketId']} created. We\'ll reply soon.';
  }
}

File: implementations/react/src/patterns/behavioral/template/concrete-support.ts
import { FormTemplate } from "./abstract-class";
/**
 * Concrete Class que implementa los pasos específicos para un flujo de contacto de soporte.
 */
export class ContactFlow extends FormTemplate {
  protected validate(data: Record<string, any>) {
    if (!data.subject) throw new Error("Subject is required");
    if (!data.message) throw new Error("Message is required");
    return {
      subject: String(data.subject).trim(),
      message: String(data.message).trim(),
      priority: data.priority || "normal",
    };
  }

  protected async send(cleanData: Record<string, any>) {
    console.log("[Contact] sending ticket:", cleanData);

    // pretend backend returns ticket info
    return {
      ticketId: "TICKET-" + Date.now(),
      status: "open",
    };
  }

  protected async afterSuccess(result: any) {
    console.log("[Contact] notify support about", result.ticketId);
  }

  protected successMessage(result: any) {
    return `Ticket ${result.ticketId} created. We'll reply soon.`;
  }
}
